### DISCUSSION TITLE
Telegram as OAuth providers

### DISCUSSION THREAD

**Discussion by Vladimir-Va** [casdoor/casdoor#1358] [Ideas]
Hello is possible to add [Telegram messenger](https://web.telegram.org/) as OAuth providers?
https://core.telegram.org/api/auth#sign-insign-up

**Comment by hsluoyz** [comment]
@Vladimir-Va issue created at: https://github.com/casdoor/casdoor/issues/1363

**Comment by forestmgy** [comment]
After researching, I think it is a bit difficult for casdoor to support telegram as an OAuth provider. 
There are two reasons:
1. Telegram does not support classic OAuth login. I googled telegram's OAuth information and found two issues. 
One is to discuss adding telegram as an OAuth provider to the nextauthjs [link here.](https://github.com/nextauthjs/next-auth/issues/2406)
Another is requesting the telegram team to add support for OAuth. It can be seen that telegram itself does not support classic OAuth [link here.](https://bugs.telegram.org/c/23608)
2. Telegram supports a login method that directly adds a button on the web page, https://core.telegram.org/widgets/login
This login method is very rigid, and can only be set according to the style given by telegram, and it is impossible to set the icon style like casdoor.
To sum up, I think adding telegram to the OAuth provider is not a good option now.

**Reply by reuzurq** [reply]
>    This login method is very rigid, and can only be set according to the style given by telegram, and it is impossible to set the icon style like casdoor.

You can insert bare script without any data attributes and later use methods like: `Telegram.Login.init({"bot_id":0,"request_access":"write"});` and `Telegram.Login.open();`

**Comment by hsluoyz** [comment]
@Vladimir-Va any comment?

**Reply by Vladimir-Va** [reply]
Hello.
As an alternative, consider [Seamless Telegram Login](https://core.telegram.org/api/url-authorization) set casdoor as Link URL authorization.
User start bot / open casdoor URL buttons in inline keyboards / User start session.
